File: src/main/java/nz/co/searchwellington/controllers/admin/PublisherAutoGatherController.java
package nz.co.searchwellington.controllers.admin;

import java.util.List;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import nz.co.searchwellington.controllers.CommonModelObjectsService;
import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.Resource;
import nz.co.searchwellington.model.Website;
import nz.co.searchwellington.modification.ContentUpdateService;
import nz.co.searchwellington.repositories.HibernateResourceDAO;
import nz.co.searchwellington.urls.UrlParser;

import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.servlet.ModelAndView;

import com.google.common.collect.Lists;

@Controller
public class PublisherAutoGatherController {

    private static Logger log = Logger.getLogger(PublisherAutoGatherController.class);
    
    private AdminRequestFilter requestFilter;
    private HibernateResourceDAO resourceDAO;
    private ContentUpdateService contentUpdateService;
    private CommonModelObjectsService commonModelObjectsService;
	private UrlParser urlParser;

    public PublisherAutoGatherController() {
	}
    
	@Autowired
	public PublisherAutoGatherController(AdminRequestFilter requestFilter,
			HibernateResourceDAO resourceDAO,
			ContentUpdateService contentUpdateService,
			CommonModelObjectsService commonModelObjectsService,
			UrlParser urlParser) {
		this.requestFilter = requestFilter;
		this.resourceDAO = resourceDAO;
		this.contentUpdateService = contentUpdateService;
		this.commonModelObjectsService = commonModelObjectsService;
		this.urlParser = urlParser;
	}
    
	@RequestMapping("/admin/gather/prompt")
	public ModelAndView prompt(HttpServletRequest request, HttpServletResponse response) {        
        ModelAndView mv = new ModelAndView();
        
        mv.setViewName("autoGatherPrompt");       
        mv.addObject("heading", "Auto Gathering");
        commonModelObjectsService.populateCommonLocal(mv);
        
        requestFilter.loadAttributesOntoRequest(request);
        Website publisher = (Website) request.getAttribute("publisher");
        mv.addObject("publisher", publisher);
        
        if (publisher != null) {    
        	List<Resource> resourcesToAutoTag = Lists.newArrayList();   
        	for (Resource resource : getPossibleAutotagResources(publisher)) {
        		if (needsPublisher((Newsitem) resource, publisher)) {
        			resourcesToAutoTag.add(resource);
        		}
        	}
            mv.addObject("resources_to_tag", resourcesToAutoTag);
        }
        return mv;
    }
	
	@RequestMapping(value="/admin/gather/apply", method=RequestMethod.POST)
	public ModelAndView apply(HttpServletRequest request, HttpServletResponse response) {        
        ModelAndView mv = new ModelAndView();
        mv.setViewName("autoGatherApply");      
        mv.addObject("heading", "Auto Gathering");
        commonModelObjectsService.populateCommonLocal(mv);
        
        requestFilter.loadAttributesOntoRequest(request);
        Website publisher = (Website) request.getAttribute("publisher");
        mv.addObject("publisher", publisher);
                
        if (publisher != null) {                       
            List<Resource> resourcesAutoTagged = Lists.newArrayList();            
            String[] autotaggedResourceIds = request.getParameterValues("autotag");            
            for (String resourceIdString : autotaggedResourceIds) {
                int resourceId = Integer.parseInt(resourceIdString);
                Resource resource = resourceDAO.loadResourceById(resourceId);

                if (resource.getType().equals("N")) {
                	log.info("Applying publisher " + publisher.getName() + " to:" + resource.getName());
                	((Newsitem) resource).setPublisher(publisher);
                	contentUpdateService.update(resource);
                	resourcesAutoTagged.add(resource);
                }
            }
            mv.addObject("resources_to_tag", resourcesAutoTagged);
        }     
        return mv;
    }
	
    private List<Resource> getPossibleAutotagResources(Resource publisher) {
		final String publishersUrlStem = urlParser.extractHostnameFrom(publisher.getUrl());
		return resourceDAO.getNewsitemsMatchingStem(publishersUrlStem);
	}
    
    private boolean needsPublisher(Newsitem resource, Website proposedPublisher) {
    	if (resource.getPublisher() == null) {
    		return true;
    	}
    	if (resource.getPublisher() != proposedPublisher) {
    		return true;    		
    	}
		return false;
	}
    
}

File: src/main/scala/nz/co/searchwellington/controllers/admin/PublisherAutoGatherController.scala
package nz.co.searchwellington.controllers.admin

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.controllers.{CommonModelObjectsService, LoggedInUserFilter, RequiringLoggedInUser}
import nz.co.searchwellington.filters.AdminRequestFilter
import nz.co.searchwellington.model._
import nz.co.searchwellington.modification.ContentUpdateService
import nz.co.searchwellington.repositories.ContentRetrievalService
import nz.co.searchwellington.repositories.mongo.MongoRepository
import nz.co.searchwellington.urls.UrlParser
import org.apache.commons.logging.LogFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.{GetMapping, PostMapping}
import org.springframework.web.servlet.ModelAndView

import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
import scala.concurrent.Await
import scala.concurrent.ExecutionContext.Implicits.global

@Controller class PublisherAutoGatherController @Autowired()(requestFilter: AdminRequestFilter,
                                                             mongoRepository: MongoRepository,
                                                             contentUpdateService: ContentUpdateService,
                                                             urlParser: UrlParser,
                                                             val contentRetrievalService: ContentRetrievalService,
                                                             val loggedInUserFilter: LoggedInUserFilter) extends
  ReasonableWaits with CommonModelObjectsService with RequiringLoggedInUser {

  private val log = LogFactory.getLog(classOf[PublisherAutoGatherController])

  @GetMapping(Array("/admin/gather/prompt"))  // TODO incorrect path
  def prompt(request: HttpServletRequest): ModelAndView = {
    def prompt(loggedInUser: User): ModelAndView = {
      val mv = new ModelAndView("autoGatherPrompt").
        addObject("heading", "Auto Gathering")

      requestFilter.loadAttributesOntoRequest(request)
      val publisher = request.getAttribute("publisher").asInstanceOf[Website]
      mv.addObject("publisher", publisher)
      if (publisher != null) {
        val resourcesToAutoTag = getPossibleAutotagResources(publisher).filter { resource =>
          needsPublisher(resource.asInstanceOf[Newsitem], publisher)
        }
        mv.addObject("resources_to_tag", resourcesToAutoTag)
      }
      mv
    }

    requiringAdminUser(prompt)
  }

  @PostMapping(Array("/admin/gather/apply"))
  def apply(request: HttpServletRequest, response: HttpServletResponse): ModelAndView = {
    def apply(loggedInUser: User): ModelAndView = {
      val mv = new ModelAndView("autoGatherApply").
        addObject("heading", "Auto Gathering")

      requestFilter.loadAttributesOntoRequest(request)
      val publisher = request.getAttribute("publisher").asInstanceOf[Website]
      mv.addObject("publisher", publisher)
      if (publisher != null) {
        val autotaggedResourceIds = request.getParameterValues("autotag") // TODO parameter name
        val resources = autotaggedResourceIds.flatMap(id => Await.result(mongoRepository.getResourceById(id), TenSeconds))
        val autotaggedNewsitems = resources.filter(resource => resource.`type` == "N").map { newsitem =>
          log.info("Applying publisher " + publisher.title + " to:" + newsitem.title)
          newsitem match {
            case p: PublishedResource => p.setPublisher(publisher)
            case _ =>
          }
          contentUpdateService.update(newsitem)
          newsitem
        }

        mv.addObject("resources_to_tag", autotaggedNewsitems)
      }

      mv
    }

    requiringAdminUser(apply)
  }

  private def getPossibleAutotagResources(publisher: Resource): Seq[Resource] = {
    val publishersHostname = urlParser.extractHostnameFrom(publisher.page)
    Await.result(mongoRepository.getNewsitemsMatchingHostname(publishersHostname), TenSeconds)
  }

  private def needsPublisher(resource: Newsitem, proposedPublisher: Website): Boolean = {
    // Apply the new publisher if the resource currently has no publisher or a different publisher
    !resource.publisher.contains(proposedPublisher._id) // TODO inline this into the query
  }

}
